FactCache: legacy update: first_order_merge not defined

##### SUMMARY
Legacy calls of FactCache.update() should be handled by FactCache.first_order_merge(), but the latter one is not defined.


##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
FactCache

##### ANSIBLE VERSION
```paste below
ansible 2.9.0.dev0 config file = None configured module search path =
['~/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
ansible python module location = ~/git/ansible/lib/ansible executable
location = ~/git/ansible/bin/ansible python version = 3.6.7 (default,
Oct 22 2018, 11:32:17) [GCC 8.2.0]

Commit 34e9c0f3a5c6
```

##### CONFIGURATION
```paste below
ansible-config dump --only-changed
```

##### OS / ENVIRONMENT

Not really relevant, Ubuntu 18.04


##### STEPS TO REPRODUCE

See #55739. I extended the tests to show the issue.

##### EXPECTED RESULTS

FactCache updating, even if legacy version of update() call is used.


##### ACTUAL RESULTS
```paste below
E           AttributeError: 'FactCache' object has no attribute 'first_order_merge'
```

--- a/changelogs/fragments/55781-fact-cache-first_merge_order.yml
+++ b/changelogs/fragments/55781-fact-cache-first_merge_order.yml
@@ -1,0 +1,2 @@
+bugfixes:
+  - fact_cache - Define the first_order_merge method for the legacy FactCache.update(key, value).

--- a/lib/ansible/vars/fact_cache.py
+++ b/lib/ansible/vars/fact_cache.py
@@ -58,6 +58,19 @@ class FactCache(MutableMapping):
         """ Flush the fact cache of all keys. """
         self._plugin.flush()
 
+    def first_order_merge(self, key, value):
+        host_facts = {key: value}
+
+        try:
+            host_cache = self._plugin.get(key)
+            if host_cache:
+                host_cache.update(value)
+                host_facts[key] = host_cache
+        except KeyError:
+            pass
+
+        super(FactCache, self).update(host_facts)
+
     def update(self, *args):
         """
         Backwards compat shim

--- a/test/units/plugins/cache/test_cache.py
+++ b/test/units/plugins/cache/test_cache.py
@@ -123,6 +123,7 @@ class TestFactCache(unittest.TestCase):
         assert self.cache['cache_key']['key'] == 'updatedvalue'
         assert self.cache['cache_key']['key2'] == 'value2'
 
+
 class TestAbstractClass(unittest.TestCase):
 
     def setUp(self):

--- a/test/units/plugins/cache/test_cache.py
+++ b/test/units/plugins/cache/test_cache.py
@@ -117,6 +117,11 @@ class TestFactCache(unittest.TestCase):
                                     "Unable to load the facts cache plugin.*json.*",
                                     FactCache)
 
+    def test_update_legacy_key_exists(self):
+        self.cache['cache_key'] = {'key': 'value', 'key2': 'value2'}
+        self.cache.update('cache_key', {'key': 'updatedvalue'})
+        assert self.cache['cache_key']['key'] == 'updatedvalue'
+        assert self.cache['cache_key']['key2'] == 'value2'
 
 class TestAbstractClass(unittest.TestCase):
 

